### DISCUSSION TITLE
Relationships not updating.

### DISCUSSION THREAD

**Discussion by jringeisen** [livewire/livewire#4766] [1. Help] [answered]
I was hoping someone can help me out here. When I use the relationship property (`auth()->user()->contentPillars`) the test fails, but if use the relationship method (`auth()->user()->contentPillars()->get()`) the test passes. What am I missing?

```php
/**
* @test
* @group Feature
*/
public function content_pillars_are_updated_when_a_new_one_is_created()
{
    $user = User::factory()->create();

    ContentPillar::factory()->create(['user_id' => $user->id]);

    $instance = Livewire::actingAs($user)
          ->test(Index::class)
          ->assertCount('contentPillars', 1);

    ContentPillar::factory()->create(['user_id' => $user->id]);

    dd(auth()->user()->contentPillars->count()); // count = 1
    dd(auth()->user()->contentPillars()->get()->count()); // count = 2

    $instance->emit('createdContentPillar')
         // Test fails here!!
          ->assertCount('contentPillars', 2);
}
```

Index::class
```php
    /**
     * @var string[]
     */
    protected $listeners = ['createdContentPillar'];

    /**
     * @return void
     */
    public function createdContentPillar()
    {
        $this->mount();
    }

    /**
     * @return void
     */
    public function mount()
    {
        $this->fill([
            'contentPillars' => request()->user()->contentPillars->pluck('name', 'id'),
            'categories' => request()->user()->categories->pluck('name', 'id'),
        ]);
    }
```

**Comment by joshhanley** [comment]
@jringeisen not related to your issue, but don't call mount like that as it's controlled by Livewire internally. I would swap them around, like this

```php
    /**
     * @var string[]
     */
    protected $listeners = ['createdContentPillar'];

    /**
     * @return void
     */
    public function mount()
    {
        $this->createdContentPillar();
    }

    /**
     * @return void
     */
    public function createdContentPillar()
    {
        $this->fill([
            'contentPillars' => request()->user()->contentPillars->pluck('name', 'id'),
            'categories' => request()->user()->categories->pluck('name', 'id'),
        ]);
    }
```

In relation to your issue, maybe the relation isn't loaded on the user model. You could try putting `auth()->user()->loadMissing('contentPillars');` before your fill function to make sure the relation is already loaded.

Hope this helps!

**Reply by jringeisen** [reply]
@joshhanley thanks for the tip! That makes sense. As for the relationship issue. Your suggestion doesn't work. I know the code works because I can use the functionality without an issue. For some reason within the test, the relationship doesn't get updated when I create the 2nd one and call the event. Also, any idea why these would return two different counts. Aren't they essentially the same thing?

```php
// dd(auth()->user()->contentPillars->count()); // count = 1
// dd(auth()->user()->contentPillars()->get()->count()); // count = 2
```

**Reply by joshhanley** [reply]
@jringeisen no worries!

Ah bugger. Yeah not sure why that wouldn't be working then, nor why it would be returning different counts.

The only thing that would cause different counts is the first option `->contentPillars` once loaded will remain in memory, so if more get added, then you will still have the old amount in memory.

Where as using `->contentPillars()->get()` it will execute a fresh query from the database every time that is called.

**Reply by jringeisen** [reply]
@joshhanley that's interesting, I didn't realize the first one is stored in memory. Do you know of a solution to get around this in tests?

**Reply by joshhanley** [reply]
@jringeisen yeah, it's called eager loading, to prevent multiple database queries https://laravel.com/docs/9.x/eloquent-relationships#eager-loading

I think there may be a way to refresh an eager loaded relation. I can't remember off the top of my head though, but just search for it, it's a pretty common thing in Laravel.

**Comment by jringeisen** [comment, marked as answer]
If I change this:
```php
$this->fill([
    'contentPillars' => request()->user()->contentPillars->pluck('name', 'id'),
    'categories' => request()->user()->categories->pluck('name', 'id'),
]);
```

to this:
```php
$this->fill([
    'contentPillars' => request()->user()->contentPillars()->get()->pluck('name', 'id'),
    'categories' => request()->user()->categories->pluck('name', 'id'),
]);
```
The test passes and the functionality still works. I hate it when I don't understand why something works the way it does lol.
